src/path.cr

```diff
@@ -0,0 +1,847 @@
+  #
+  # See also Rob Pike: *[Lexical File Names in Plan 9 or Getting Dot-Dot Right](https://9p.io/sys/doc/lexnames.html)*
+  def normalize(*, remove_final_separator : Bool = true) : Path
+    return new_instance "." if @name.empty?
+
+    len = @name.size
+    rooted = absolute? || (windows? && @type.starts_with_separator?(@name))
+    reader = Char::Reader.new(@name)
+    dotdot = 0
+    separator_ord = @type.separators.map &.ord.to_u8
+    add_separator_at_end = !remove_final_separator && @type.ends_with_separator?(@name)
+
+    new_name = String.build do |str|
+      if rooted
+        str << @type.separator
+        reader.next_char
+        dotdot = 1
+      end
+
+      while (char = reader.current_char) != Char::ZERO
+        curr_pos = reader.pos
+        if @type.separators.includes?(char)
+          # empty path element
+          reader.next_char
+        elsif char == '.' && (reader.pos + 1 == @name.bytesize || @type.separators.includes?(reader.peek_next_char))
+          # . element
+          reader.next_char
+        elsif char == '.' && reader.next_char == '.' && (reader.pos + 1 == @name.bytesize || @type.separators.includes?(reader.peek_next_char))
+          # .. element: remove to last /
+          reader.next_char
+          if str.bytesize > dotdot
+            str.back 1
+            while str.bytesize > dotdot && !separator_ord.includes?((str.buffer + str.bytesize).value)
+              str.back 1
+            end
+          elsif !rooted
+            if str.bytesize > 0
+              str << @type.separator
+            end
+            str << ".."
+            dotdot = str.bytesize
+          end
+        else
+          reader.pos = curr_pos # make sure to reset lookahead used in previous condition
+
+          # real path element
+          # add slash if needed
+
+          if rooted && str.bytesize != 1 || !rooted && str.bytesize != 0
+            str << @type.separator
+          end
+
+          loop do
+            str << char
+            char = reader.next_char
+            break if @type.separators.includes?(char) || char == Char::ZERO
+          end
+        end
+      end
+
+      if str.empty?
+        str << '.'
+      end
+
+      last_char = (str.buffer + str.bytesize - 1).value.unsafe_chr
+
+      if add_separator_at_end && !@type.separators.includes?(last_char)
+        str << @type.separator
+      end
+    end
+
+    new_instance new_name
+  end
+
+  # Yields each component of this path as a `String`.
+  def each_part
+    @type.each_part(@name) do |part|
+      yield part
+    end
+  end
+
+  # Returns the components of this path as an `Array(String)`.
+  def parts : Array(String)
+    parts = [] of String
+    each_part do |part|
+      parts << part
+    end
+    parts
+  end
+
+  # Converts this path to a Windows path.
+  def to_windows : Path
+    return self if windows?
+
+    new_instance(@name, Type::WINDOWS)
```

Comment:
Even if Windows supports `/` isn't `\` the default? What do Windows API functions return for example?

Reply:
The path `foo/bar` has the same interpretation as windows and posix path. Converting from one to the other shouldn't change the string representation. If you want to change all forward slash directory separators to back slashes, you can use `to_windows.normalize`. I don't think the converter method should automatically normalize all paths. That might not be intended.